Compare additionally generation when comparing ImageStreamTag status with spec

soltysh opened the pull request:
Fixes #10402.

The root cause in the linked issue was that there is no actual change in image ID or it's reference and thus this change was not caught when comparing. I've added comparing generation in that code, as well.

@smarterclayton ptal
@mfojtik fyi

smarterclayton commented on pkg/image/api/helper.go:
```diff
@@ -642,7 +642,8 @@ func DifferentTagEvent(stream *ImageStream, tag string, next TagEvent) bool {
 	sameImage := previous.Image == next.Image
-	return !(sameRef && sameImage)
+	sameGeneration := previous.Generation == next.Generation
+	return !(sameRef && sameImage && sameGeneration)
```
See the Godoc on this method.  If there's a reason why the generation is not compared, we need to handle that as well.

soltysh replied:
You're suggesting to add a flag, or create additional method?

smarterclayton replied:
Not sure - there was a reason not to mutate the generation, so knowing which callers depended on it and which use cases would help clarify.  I'm scared there aren't enough tests to be sure.

soltysh replied:
I'll add the parameter and make sure to cover all, or at least most of the paths with new tests.

soltysh commented:
Since there's only call of that method I'd like to see how the tests are working. In the mean time I'm adding more unit tests.

[test]

soltysh commented:
@smarterclayton I've updated the patch to compare generations between status and spec not from an tagevent which is meant to be the feature tag and added a bit of tests. ptal

smarterclayton commented:
The change looks good, if you get green I'll flag it for merge.

mfojtik commented on pkg/image/api/helper.go:
```diff
@@ -645,6 +645,18 @@ func DifferentTagEvent(stream *ImageStream, tag string, next TagEvent) bool {
 	return !(sameRef && sameImage)
 }
 
+func DifferentTagGeneration(stream *ImageStream, tag string) bool {
```
need godoc

mfojtik commented:
LGTM

please add godoc and @smarterclayton can tag it for merge

soltysh commented:
Tests are (were) green, addressed Michal comment about missing godoc. @smarterclayton this is good to merge in that case.

smarterclayton commented:
[merge]

smarterclayton commented:
[test]

soltysh commented:
There are a few errors, but it all looks like caused by devicemapper error flake from #9548.

[test]

smarterclayton commented:
[merge]
